ad9656:Add driver for ad9656 chip

hotoleanudan opened the pull request:
Added driver for the ad9656 chip.

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+{
+	int32_t ret;
+
+	ret = spi_remove(dev->spi_desc);
```
check for return

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+			 0x3C);	// RESET
+	mdelay(250);
+
+	ad9656_spi_write(dev,
```
this can be fit in one line. The same should be applied for the other cases

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+	ad9656_spi_read(dev,
+			AD9656_REG_CHIP_ID,
+			&chip_id);
+	if(chip_id != AD9656_CHIP_ID) {
```
Linux coding style:
```
if (condition) not 
if(condition)
```

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+	/* SPI */
+	ret = spi_init(&dev->spi_desc, &init_param->spi_init);
+
+	ad9656_spi_read(dev,
```
check read operation
```
ret = ad9656_spi_read(dev,
if (ret < 0)
            return ret;
```

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+	int32_t ret;
+	struct ad9656_dev *dev;
+
+	ret = 0;
```
no need to initialize here if ret is used in read/write operations checks.

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+	ad9656_spi_write(dev,
+			 AD9656_REG_ADC_TEST_MODE,
+			 test_mode);
+	if (test_mode == AD9656_TEST_OFF)
```
this can be restructure to 
```
if (test_mode == AD9656_TEST_OFF)
         tmp = AD9656_FORMAT_2S_COMPLEMENT;
else
         tmp = AD9656_FORMAT_OFFSET_BINARY;

return ad9656_spi_write(dev,  AD9656_REG_OUTPUT_MODE, tmp);

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+
+	int32_t ret;
+
+	// the MSB of byte 0 indicates a r/w operation, following 7 bits are the 
```
this is the style of comments: https://www.kernel.org/doc/html/v4.10/process/coding-style.html#commenting
Also use standard C comment for one-liners
```
/* Comment */

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+ * @brief Implementation of AD9656 Driver.
+ * @author DHotolea (dan.hotoleanu@analog.com)
+ ********************************************************************************
+ * Copyright 2014-2016(c) Analog Devices, Inc.
```
copyright 2020

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+
+	ret = spi_write_and_read(dev->spi_desc,
+				 buf,
+				 3);
```
check function return, maybe the value for buf[2] is invalid no need to assign it to reg_data in this case

hotoleanudan replied:
I don't see how this can be checked because you are reading a register that can have any value.

mirceacaprioru replied:
I am referring to the function return. 
```
ret = spi_write_and_read()
if (ret < 0)
        return ret;

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+			uint8_t *reg_data)
+{
+	uint8_t buf[3];
+
```
nitpick: remove extra line

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+	buf[1] = reg_addr & 0xFF;
+	buf[2] = reg_data;
+
+	ret = spi_write_and_read(dev->spi_desc,
```
return directly:
```
return pi_write_and_read(dev->spi_desc,

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.h:
```diff
@@ -0,0 +1,104 @@
+ * @brief Header file of AD9656 Driver.
+ * @author DHotolea (dan.hotoleanu@analog.com)
+ ********************************************************************************
+ * Copyright 2014-2016(c) Analog Devices, Inc.
```
copyright 2020

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+	buf[1] = reg_addr & 0xFF;
+	buf[2] = 0x00;
+
+	ret = spi_write_and_read(dev->spi_desc,
```
this can be fitted in one line.

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+	// bits 14-8 of the address of the register that is accessed. Byte 1 
+	// contains the bits 7-0 of the address of the register. Byte 2 contains
+	// the data to be written.
+	buf[0] = reg_addr >> 8;
```
The code duplication here is just 2 lines. I don't think creating another wrapper function for this will improve the code. It would just be used in spi_read and spi_write functions.

mirceacaprioru replied:
For readability and code understanding i would leave it as is.

adrimbarean commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,217 @@
+/***************************************************************************//**
```
Commit needs Signed-off-by tag.

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,202 @@
+/***************************************************************************//**
```
The modifications that were done must be included in a changelog in the pull request something like:
```
Changelog v2:
- added Signed-off-by tag
- check return value
```
The commit description must contain in this case some details of the chip and some capabilities that the driver supports.

mirceacaprioru replied:
The commit still does not describe nothing about the chip. 
```
Modified the code according to the pull request comments.
```
Is not telling any information about the implementation and has no value.

hotoleanudan replied:
Changelog v2:

- checked the return value of the spi_remove operation
- modified the call for the functions to be written in one line
- modified if clause to respect the Linux coding style
- deleted the unnecessary initialization of return variable
- restructured the code for the JESD204 test function
- modified the commenting style from // to /* */
- modified the copyright year
- checked if the read SPI function was successful
- deleted unnecessary return variable and called the spi_write_and_read directly from return
- used struct clause when declaring a spi_init_param variable
- used struct clause when declaring a spi_desc variable
- replaced the include of "platform_drivers.h" with "spi.h"
- modified and completed the function comments
- checked return value for spi_init function

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,202 @@
+
+/**
+ * @brief Launch the test functionality for the receive path for the ad9656 chip
+ * @param dev - The device hanlder for the ad9656 chip
```
there is a small typo here  "handler"

hotoleanudan replied:
Done

adrimbarean commented:
Please apply AStyle to the files. Also the commit message needs to describe what it is doing in the commit and not refer to the pull request.

hotoleanudan commented:
Can you please tell me if there are any more things to modify to this driver? If not can I merge it to master?

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -108,25 +108,51 @@ int32_t ad9656_spi_write(struct ad9656_dev *dev,
-	else
 		format = AD9656_FORMAT_OFFSET_BINARY;
+	else
+		format = AD9656_FORMAT_2S_COMPLEMENT;
```
nitpick: add extra line before return.

hotoleanudan replied:
Done

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -150,15 +176,15 @@ int32_t ad9656_setup(struct ad9656_dev **device,
 	if (ret != SUCCESS)
 		return ret;
 
+	ad9656_spi_write(dev, AD9656_SPI_CONFIG, 0x3C);	/* RESET */
```
modification not related to commit, probably it should go in the first one.

hotoleanudan replied:
I just moved the reset instruction for the AD9656 chip to be the first instruction that is executed on this chip in the setup process

mirceacaprioru commented:
Don't forget about the changelog describing the modifications done after a force push. That will send an email and inform us that something has been modified in this PR, and is also valuable to determine what issues have been addressed from the previous review.

hotoleanudan commented:
Changelog v3:

- applied AStyle on the driver code
- checked the return value for the ad9656_spi_read function inside the ad9656 setup function
- inserted new line in the JESD204 testing function before the return call

adrimbarean commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,202 @@
+ * @param reg_data - The value read from the internal register
+ * @return SUCCESS if the value was successfully read, FAILURE otherwise
+ */
+int32_t ad9656_spi_read(struct ad9656_dev *dev,
```
I would rename those `ad9656_reg_read` and `ad9656_reg_write` to be consistent with other drivers.

hotoleanudan replied:
Done in v5

adrimbarean commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -116,21 +116,22 @@ int32_t ad9656_spi_write(struct ad9656_dev *dev,
  * 		   FAILURE otherwise
```
Title of the commit message is a bit too long.

hotoleanudan replied:
Squashed the commit to the previous commits

hotoleanudan commented:
Changelog v4:

- Re-based the pr such that all commits for this pr that were pushed on the server without amending the first commit and force pushing where squashed in the last amended commit that was made. The squashed changelog for these commits is:
        - In ad9656.c checked if the return values from function calls are correct.
        - In ad9656.c and ad9656.h modified the copyright year of the file description.
        - In ad9656.c modified the function calls that were written on multiple lines to be written on a single line.
        - In ad9656.c modified code to comply with Linux coding style.
        - In ad9656.c deleted unneeded variable initializations.
        - In ad9656.c rearranged the code in the ad9656_JESD204_test function.
        - In ad9656.c modified the single line commenting style from // to /**/.
        - In ad9656.c deleted unused return variable.
        - In ad9656.h used "struct" type declaration for the declaration of variables of type spi_init_param and spi_desc.
        - In ad9656.c swapped the inclusion of "platform_drivers.h" with "spi.h".
        - In ad9656.c wrote function comments for each function in the file.
        - In ad9656.c corrected typos.
        - In ad9656.c and ad9656.h added function for generating user test pattern sequence for the ADC.
        - Ran AStyle software on both ad9656.c and ad9656.h files.

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,248 @@
+/***************************************************************************//**
```
the changelog was added in the commit description. The commit description should contain some information about the driver some basic info about the part not fixes that were done after the review process.

hotoleanudan replied:
Changed the commit description to contain a brief description of the AD9656 chip.

hotoleanudan commented:
Changelog v5:

- Renamed the ad9656_spi_read and d9656_spi_write functions to ad9656_reg_read and ad9656_reg_write.
- Changed the commit comment to a brief description of the AD9656 chip.

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,248 @@
+	uint8_t format;
+
+	/* write user input test pattern in the registers */
+	ad9656_spi_write(dev, AD9656_REG_USER_TEST_PATTERN_1_LSB,
```
check return for each spi_write. If something fails it make no sense to go further.

hotoleanudan replied:
Done in v6

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,248 @@
+	int32_t ret;
+	struct ad9656_dev *dev;
+
+	dev = (struct ad9656_dev *)malloc(sizeof(*dev));
```
calloc would also set the memory to 0 to be sure that there is no residual values left over in memory.

hotoleanudan replied:
Done in v6

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,248 @@
+	if (ret != SUCCESS)
+		return ret;
+
+	ad9656_spi_write(dev, AD9656_SPI_CONFIG, 0x3C);	/* RESET */
```
check spi_write return value.

hotoleanudan replied:
Done in v6

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,248 @@
+		return FAILURE;
+	}
+
+	ad9656_spi_write(dev, AD9656_REG_LINK_CONTROL,
```
check return value, do this for all other occurrences of spi_write.

hotoleanudan replied:
Done in v6

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,248 @@
+	ad9656_spi_write(dev, AD9656_REG_JESD204B_QUICK_CONFIG, 0x44);	/* m=4, l=4 */
+	ad9656_spi_write(dev, AD9656_REG_JESD204B_SCR_L,
+			 0x83);	/* enable scrambling, l=4 */
+	if (init_param->lane_rate_kbps < 2000000)
```
this can be restructured by moving the spi_write outside of the if statement and using a tmp variable to enable or disable the line rate. Remeber to check the return value of spi_write

hotoleanudan replied:
Done in v6

hotoleanudan commented:
Changelog v6:

- Changed all occurences of ad9656_spi_read and ad9656_spi_write to ad9656_reg_read and ad9656_reg_write.
- Checked the return types of all the ad9656_reg_write functions.
- Changed the memory allocation function of the ad9656_dev variable in ad9656_setup from malloc to calloc.
- Extracted the ad9656_reg_write function call from the if clause that checks the lane rate speed and placed it after the if clause. For this I used a temporary variable that I assigned the value of the lane rate control register.

hotoleanudan commented:
Changelog v7:

- Applied AStyle to c file

mirceacaprioru commented on drivers/adc/ad9656/ad9656.c:
```diff
@@ -0,0 +1,313 @@
+
+	ret = ad9656_reg_write(dev, AD9656_REG_LINK_CONTROL,
+			       0x15);	/* disable link, ilas enable */
+
```
no need for extra line here. These two go together without an empty line separating them.

hotoleanudan replied:
Deleted all the extralines in v8

hotoleanudan commented:
Changelog v8:

- Deleted all extralines in c file

mirceacaprioru commented:
LGTM one thing i would like to get some feedback is the return messages. SUCCESS FAILURE, are we still using this or are we shifting to errno.h messager ? @buha @mchindri @dbogdan 

If there are no remakrs will merge later today.

hotoleanudan commented:
Changelog v9:

- re-based this branch because it was behind the master branch.
